### DISCUSSION TITLE
Are we all reading from the same page?

### DISCUSSION THREAD

**Discussion by fade2gray** [mmikeww/AHK-v2-script-converter#234] [Ideas]
Can some ground rules be laid?

Literal tabs or spaces for indentation and alignment - if spaces, how many - 2/3/4, all have been used? I notice a vast mix of both used throughout the code, even mixing literal tabs and spaces in the same indentation or alignment spacing,

Other stuff that we should be reading from the same page?

**Comment by andymbody** [comment]
I've noticed the same and it's a lot more work to make the code neat when spaces are used instead of tabs. With spaces, each time a backspace is required, there are multiple keystrokes involved, and then multiple keystrokes involved again just to get the code to line back up. @mmikeww has made it clear that spaces are not to be changed to tabs. I don't know the reason for this, as tabs are much less work and easier to manage (IMO) than multiple spaces covering the same real-estate within the document.

If a vote takes place for a unified formatting, my vote  will be cast for tabs4. But obviously, I will have to live with the general consensus.

**Reply by fade2gray** [reply]
> With spaces, each time a backspace is required ...

It's not, at least not with VSCode - just set the number of spaces required for your tabs, and a single backspace will remove however many spaces used for each tab.


https://github.com/mmikeww/AHK-v2-script-converter/assets/1332009/8ba68649-71cc-4873-9e73-d79267143a23

**Reply by andymbody** [reply]
> just set the number of spaces required for your tabs, and a single backspace will remove however many spaces used for each tab.

This works fine for leading tabs, but not trailing tabs (between code and trailing line comments - at least in Scite). I won't be switching to VS just for this minor improvement. I'm using Win7, and I doubt its supported anyway. But thanks for the heads up.

**Comment by fade2gray** [comment]
> but not trailing tabs (between code and trailing line comments

True, but you could set up a hotkey to backspace x4. 😜

**Comment by mmikeww** [comment]
spaces vs tabs debate has existed for as long as people have ever started coding

i'm a spaces person, and so to keep the project consistent, thats what i originally chose

but obviously as more contributors join, code can get messy and mixed. ultimately for a real solution, an editorconfig file should be used, and most programmers should be using a text editor that supports an editorconfig file, and then all of the issues are likely erased. the editor handles it on a per-project basis, depending on what the editorconfig file says

https://editorconfig.org/

so the necessary steps would be

1. vote on what standards we should use. you guys are active contributors so your opinion is more important. i'm not actively working on the project at the moment
2. someone sets up an editorconfig file for the project and makes a PR to merge it into the repo. something like this:
```
root = true

[*]
end_of_line = crlf
insert_final_newline = true

[*.ahk]
indent_style = space
indent_size = 4
```
3. all contributors set up their text editors and whatever editorconfig plugin might be necessary to ensure that their editor pickup the editorconfig file and change the settings appropriately when working on this project

**Reply by Banaanae** [reply]
> insert_final_newline = true

What about tests? Sometimes the test suite trims the trailing newlines, other times it adds them?

Otherwise I'm happy with any standards

**Reply by Banaanae** [reply]
Did a bit of research, I suggest this

```
root = true

[*]
insert_final_newline = true
end_of_line = crlf


[*.ahk]
indent_style = space
indent_size = 4

[*.{ah1,ah2}]
# Don't enforce newlines as this can make tests fail
insert_final_newline = false
```

**Reply by Banaanae** [reply]
If we do add this:
@fade2gray https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig
@andymbody See editor.config.enable https://www.scintilla.org/SciTEDoc.html

**Reply by mmikeww** [reply]
> What about tests? Sometimes the test suite trims the trailing newlines, other times it adds them?

good point, then i guess that can be omitted

**Reply by fade2gray** [reply]
@Banaanae 
Please explain ...
```
[*]
insert_final_newline = true
end_of_line = crlf
insert_final_newline = true
```
2x `insert_final_newline = true`?

`insert_final_newline = unset` rather than `insert_final_newline = false`?
https://spec.editorconfig.org/
![EditorConfig](https://github.com/mmikeww/AHK-v2-script-converter/assets/1332009/0824a111-6cc4-4fe7-8c55-83d5b1a60fc3)

Any concerns with unresolved issue - is it just an issue for VSCode and not for VSCodium?
https://github.com/editorconfig/editorconfig-vscode/issues/349

**Reply by Banaanae** [reply]
I'll need to test to see if that affects us (doesn't look good D: )
Maybe we just don't specify it at all

**Reply by andymbody** [reply]
> What about tests? Sometimes the test suite trims the trailing newlines, other times it adds them?

My last pull request should have corrected this inconsistency. There were at least 4 functions that were causing this issue, but should now be corrected. Going forward, the converted code should not add or subtract CRLFs that were not present in the original.. With this issue resolved, the formatting can now be applied with predictable results.

**Comment by mmikeww** [comment]
`false` might explicitly REMOVE the ending newline char, whereas `unset` might just do nothing. would need to test

**Comment by fade2gray** [comment]
Thoughts?

```Ruby
/*******************************************************************
 * Replaces 3 space tabs with 4 space tabs in leading indentation. *
 * Also replaces all literal tab stops with 4 space tabs.          *
 * Interspacing needs to be tidied up manually.                    *
 *******************************************************************/

#Requires AutoHotkey v2.0

filename_1 := "ConvertFuncs.ahk"
filename_2 := "ConvertFuncs_2.ahk"
newContent := ""
oldTabSpaces := "   "
newTabSpaces := "    "

loop read filename_1
{
    newLeadingTabs := ""
    RegExMatch(A_LoopReadLine,"^(" oldTabSpaces ")+(?! )", &leadingWhitespace)
    if leadingWhitespace
    {
        leadingWhitespaceLen := leadingWhitespace.Len
        oldLeadingTabsCount := Round(leadingWhitespaceLen / 3)
        loop oldLeadingTabsCount
            newLeadingTabs .= newTabSpaces
    }
    newLine := RegExReplace(A_LoopReadLine, "^(   )+(?! )(.*)", newLeadingTabs . "$2")
    newContent .= newLine "`r`n"
}
newContent := RegExReplace(newContent,"\t", newTabSpaces)
newContent := RegExReplace(newContent,"\s*$","")

If FileExist(filename_2)
    FileDelete(filename_2)
FileAppend newContent, filename_2

ExitApp
```

**Reply by andymbody** [reply]
> Thoughts?

I can look into the options for formatting the code so that it arrives on github in a format that is agreed upon by all here. It looks like spaces get the majority vote. I think the manner in which the code arrives is the important thing, but which process we choose to get it in that format should be left up to the individual. A mass conversion before/after is one option, and this approach provides the freedom to edit the code in a format that the individual is most comfortable with, yet still meet the requirements of the project/team. The editing takes the most time, so I think this freedom should be left intact. Mass conversion takes less than a second once the routines are written. This is probably the approach I am most comfortable with, so your code suggestion has merit as far as I'm concerned.

**Reply by Banaanae** [reply]
Once editorconfig is set opening the file in any ide should automatically update it. If not then we can use this

**Comment by andymbody** [comment]
On a related note...

I'd like to say that I would be in favor of adding comments to code that is edited giving a bit of details regarding the edit's purpose and make notes that may be helpful to others reading the code. For instance, when we add or change regex needles, it would be nice to have a brief description about what that needle is extracting/replacing. Without this, I (for one) must read the needle from left to right very slowly to see what it is doing. A note would allow the reader to see the intention and only be required to dig into the needle details if there seems to be a problem with the output. That's just one example.

Comments do increase file size but also increase clarity for others who need to find a place to insert their own code. The comments I add to my code help me tremendously to remember my thought process at the time the code was written. I'm hoping it is also helpful to others who read that code. This is the type of thing I have in mind.

Anyone else in favor of this idea? Not necessarily for themselves alone, but also for the "team" as a whole.

**Reply by Banaanae** [reply]
I always add comments for an entire code block, or for when I feel that a section of code is unclear.

For RegExs (especially the big ones) I always save it in RegEx101 and share link, on that site you can also share what the RegEx matches and what is doesn't (also has pcre1 support).

When it comes to file size I'd don't really see much of a problem, documenting code is always important, and if I can figure out how to compile ahks automatically (I'm this close I just need to get the compiled ahk and release it) the comment will get stripped anyway
